### PR TITLE
mesa: restore OpenCL support.

### PR DESCRIPTION
**Pull request by sbruder**
Continuation of #82729.

Works with a Radeon RX 480 on NixOS 21.05, see https://github.com/NixOS/nixpkgs/pull/82729#issuecomment-910303664 for details.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix
Comment (rnhmjoj) [inline pkgs/top-level/all-packages.nix:17769]: Considering there's already a `mesa_glu`, it should probably be named `mesa_opencl` for consistency.

File: pkgs/development/libraries/mesa/default.nix
Comment (primeos) [inline pkgs/development/libraries/mesa/default.nix:138]: I think we should only add them with `enableOpenCL` enabled for now (to avoid accidentally depending on them after Mesa updates and to "document" this).

File: pkgs/development/libraries/mesa/default.nix
Comment (primeos) [inline pkgs/development/libraries/mesa/default.nix:130]: Would be nice if we could briefly document such options. Maybe something like "Enable the gallium OpenCL frontend"?

Description:
```
option(
  'gallium-opencl',
  type : 'combo',
  choices : ['icd', 'standalone', 'disabled'],
  value : 'disabled',
  description : 'build gallium "clover" OpenCL frontend.',
)

```

File: pkgs/development/libraries/mesa/default.nix
Comment (primeos) [inline pkgs/development/libraries/mesa/default.nix:131]: We should be able to drop this:
```
option(
  'opencl-spirv',
  type : 'boolean',
  value : false,
  description : 'build gallium "clover" OpenCL frontend with SPIR-V binary support.',
)
```

File: pkgs/development/libraries/mesa/default.nix
Comment (primeos) [inline pkgs/development/libraries/mesa/default.nix:195]: Is this still required with the `enableOpenCL` approach?
The moved files are also unconditionally referenced below.

**Review by primeos — CHANGES_REQUESTED** [review]
Overall this looks mergeable to me with a few changes. Thanks!

**Comment by kvtb** [comment]
Works on Nvidia P2000 (not exactly this PR, but my own rebase of  #82729 to the latest `mesa` which is basically the same as this PR)

File: pkgs/development/libraries/mesa/default.nix
Comment (sbruder) [reply pkgs/development/libraries/mesa/default.nix]: No, I just copied it from the version that didn’t use the enable flag without realising that.

**Comment by sbruder** [comment]
Force pushed to change the commit message (there shouldn’t be a `.` at the end).

**Comment by kvtb** [comment]
This way to enable OpenCL (taken from @sbruder  https://github.com/NixOS/nixpkgs/pull/82729#issuecomment-910303664)
```
{ pkgs, ... }:
{
  hardware.opengl =
    let
      pkgsMesaClover = import
        (pkgs.fetchFromGitHub {
          owner = "sbruder";
          repo = "nixpkgs";
          rev = "41d33334ec2d07da8f5bd0f3749949241c42266b";
          sha256 = "sha256-sA/kwZ8kXuRC/sLjja2ISpyTnlEwZjifyYeiz9hFWRE=";
        })
        { inherit (pkgs) system; };
    in
    {
      package = pkgsMesaClover.mesa-opencl.drivers;
      extraPackages = [
        pkgsMesaClover.mesa-opencl
      ];
    };
}
```
does not required mass-rebuild because of `mesa` compiled with different flags, and open a space to abandon the `enableOpelCL` flag at all.
Just to compile `mesa` always as though `enableOpelCL=true`, but copy the drivers which bring huge `clang` closure to a separate output.
And then just `hardware.opengl.package = pkgs.mesa.opencl-drivers;` to enable OpenCL (not the dot is after `mesa`, not `mesa-opencl`)
Will it work ?

**Comment by sbruder** [comment]
I’m not sure if I understand you correctly, but according to https://github.com/NixOS/nixpkgs/pull/82729#issuecomment-600062231 using a separate output for OpenCL is not easily possible.

Concerning mass rebuilds, that will only affect nixpkgs once (when this PR is merged), but after that it only requires building one additional package (`mesa_opencl`), which will be done on Hydra, so there are no mass-rebuilds necessary for users. That is because mesa is not necessarily a dependency of all packages using it (though there are exceptions, but they will still use `mesa` and not `mesa_opencl`, which should be fine). Instead, the version present at `/run/opengl-driver` is loaded dynamically because otherwise it would be impossible to use e.g. the proprietary nvidia driver without every package using hardware acceleration depending on it.

That is the case independent of what method is used to make the OpenCL support optional (separate output or different package/guards).

**Comment by kvtb** [comment]
But `mesa_opencl` contains all the files from `mesa` + open cl drivers?
And the files duplicated from `mesa` are not really needed

**Comment by kvtb** [comment]
As "mass-rebuild way" I meant the way to enable OpenCL which I used:
```nix
(super: self: {
  mesa = super.mesa.override { enableOpenCL = true; }
})
```

Here is a single `mesa`, but mass-rebuild.
In the variant in https://github.com/NixOS/nixpkgs/pull/136402#issuecomment-911717440 there is no mass-rebuild, but many duplicated files.
It seems possible to end up with a solution without mass-rebuild and duplicated files (which I thought adding an output could be)

**Comment by sbruder** [comment]
> But mesa_opencl contains all the files from mesa + open cl drivers?

Yes it does. However, duplicated files are not that uncommon on a NixOS system and are introduced with every big mass-rebuild (e.g. on an openssl (patch) update, most packages depending on it do not change, but are still stored multiple times). To only store the files once there is `nix-store --optimise` and the `auto-optimise-store` Nix option which hard-link the files so it is not that much of a problem.

While it might be cleaner to move the OpenCL support it into a separate output, it seems to be not that easily possible. Though if you manage to get it working with an extra output, I’m more than happy to use that instead.

> ```nix
> (super: self: {
>   mesa = super.mesa.override { enableOpenCL = true; }
> })
> ```

Overriding a widely-used package globally will always cause mass-rebuilds. For that reason, `hardware.opengl.packages` and `hardware.opengl.extraPackages` should be used in this case.

**Comment by kvtb** [comment]
So I suggested that the additional output of `mesa.opencl-driver` could contain files which are present in `mesa_opencl` and are not present in `mesa` (because `mesa_opencl` is supposed to be used where `mesa` is already present).
Never mind, this is just a small optimization that can be done later.
It's more important to somehow repair OpenCL, which has been broken for months.

**Comment by sbruder** [comment]
On further testing it turned out that installing OpenCL support into its own output is not that complicated. I force-pushed my attempt at doing it. Enabling OpenCL support now works by setting `hardware.opengl.extraPackages = [ pkgs.mesa.opencl ];`, the internal `package` option does not need to be modified.

File: pkgs/development/libraries/mesa/default.nix
Comment (kvtb) [inline pkgs/development/libraries/mesa/default.nix:124]: maybe opencl stuff a bit below, under `isLinux`?

Comment (sbruder) [reply pkgs/development/libraries/mesa/default.nix:120]: Good point. I moved the options to the Linux-specific ones.

File: pkgs/development/libraries/mesa/default.nix
Comment (primeos) [inline pkgs/development/libraries/mesa/default.nix:16]: Note to self: We probably want to package this as `llvmPackages*.libclc` to keep it up-to-date, etc. (outside the scope of this PR but in case anyone has an opinion on that).

File: pkgs/development/libraries/mesa/default.nix
Comment (primeos) [inline pkgs/development/libraries/mesa/default.nix:194]: Did you test if this works? I guess this gets linked as `/run/opengl-driver/etc/OpenCL/vendors/mesa.icd` but I'm not sure if that path is searched and I haven't tested it yet.

**Review by primeos — APPROVED** [review]
I didn't have time to test it yet / investigate if this affects the other outputs and their closure sizes but the diff LGTM now and enabling it by default on Linux seems reasonable (due to the separate output).

File: pkgs/development/libraries/mesa/default.nix
Comment (sbruder) [reply pkgs/development/libraries/mesa/default.nix:194]: Yes, this worked when I tested it. The [NixOS manual](https://nixos.org/manual/nixos/stable/index.html#sec-gpu-accel-opencl) mentions `hardware.opengl.extraPackages` (which get linked to `/run/opengl-driver`) as a way for installing OpenCL ICDs, so that seems to be the intended path for them.

**Review by rnhmjoj — APPROVED** [review]
I haven't tested the driver but the changes look reasonable to me.
The closure size for the other outputs didn't change:
- before:
  - out:         33.2M
  - dev:         38.4M
  - drivers:    864.5M

- after:
  - out:         33.2M
  - dev:         38.4M
  - drivers:    864.5M
  - opencl:       1.0G

so I think it's safe to be merged. Even if the drivers have some issue it's still an improvement over the current situation.

File: pkgs/development/libraries/mesa/default.nix
Comment (puchneiner) [inline pkgs/development/libraries/mesa/default.nix:17]:
```suggestion
, enableOpenCL ? stdenv.isLinux && stdenv.isx86_64
}:
```

some libclc's deps (spirv compiler, ...) are broken on non-x86_64, so mesa will be broken on non-x86_64 if we merge this in present form

**Review by puchneiner — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/mesa/default.nix
Comment (puchneiner) [reply pkgs/development/libraries/mesa/default.nix:17]: at least it breaks i686

Comment (sbruder) [reply pkgs/development/libraries/mesa/default.nix:18]: Thanks for noticing. I added that condition.

**Comment by primeos** [comment]
@ofborg test sway

**Review by primeos — APPROVED** [review]
@sbruder I'm really sorry about the long delay. It was a very busy month for me and I lost track of this PR / didn't get around to merge it. I catched a last mistake and will merge this PR once CI (hopefully) passes. Thanks a lot for this PR and your fast responses!

Also huge thanks to everyone who reviewed this!
(Especially @rnhmjoj for verifying the closure sizes.)

**Comment by primeos** [comment]
Let's ignore CI then `¯\_(ツ)_/¯` (previous results were fine; apart from timeouts).

I already ran the build and Sway test locally anyway:
```console
$ nix-build -A nixosTests.sway
/nix/store/i6c1ybq9jlbha0h6dzzac83agm4vdcwq-vm-test-run-sway
```

File: pkgs/development/libraries/mesa/default.nix
Comment (davidak) [reply pkgs/development/libraries/mesa/default.nix:17]: i suggested it before, but it seem not to be done yet

https://github.com/NixOS/nixpkgs/issues/89187#issuecomment-636227881
https://github.com/NixOS/nixpkgs/pull/111065#issuecomment-879325651

**Comment by kurnevsky** [comment]
I get the following error when trying to use opencl:
```
<built-in>:1:10: fatal error: 'clc/clc.h' file not found
```

Tried to add libclc to the system environment and to opengl.extraPackages - neither worked. Any idea how to fix it?

**Comment by kouakitaki** [comment]
I am getting the same error

```
<built-in>:1:10: fatal error: 'clc/clc.h' file not found
```

When trying to run a hello-world program for OpenCL. It seems that it tried to find that file, but the path is ill-formatted. See the repeated path `/nix/store/d902ifqzwfld7mjw3vxgy08l87055whr-libclc-11.0.1//nix/store/d902ifqzwfld7mjw3vxgy08l87055whr-libclc-11.0.1/include` in the strace below.

```
$ strace ./hello 2>&1 | rg clc
newfstatat(AT_FDCWD, "/nix/store/d902ifqzwfld7mjw3vxgy08l87055whr-libclc-11.0.1//nix/store/d902ifqzwfld7mjw3vxgy08l87055whr-libclc-11.0.1/include", 0x7ffe2a5cf060, 0) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/d902ifqzwfld7mjw3vxgy08l87055whr-libclc-11.0.1/share/clc/polaris10-amdgcn-mesa-mesa3d.bc", O_RDONLY|O_CLOEXEC) = 8
newfstatat(AT_FDCWD, "./clc", 0x7ffe2a5cde50, 0) = -1 ENOENT (No such file or directory)
write(2, "<built-in>:1:10: fatal error: 'c"..., 57<built-in>:1:10: fatal error: 'clc/clc.h' file not found
```

**Comment by shpinog** [comment]
Are there any plans to add this in the stable branch?

**Comment by vcunat** [comment]
This PR is quite old, contained in both 21.11 and 22.05 releases.

**Pull request merged**
